SYCL/ESIMD/api/functional/value_conv.hpp

```diff
@@ -0,0 +1,117 @@
+//===-- value_conv.hpp - This file provides common functions generate values for
+//      testing. ----------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file provides functions that let obtain converted reference data for
+/// test according to current underlying types.
+///
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "value.hpp"
+
+namespace esimd_test::api::functional {
+
+// Utility class to retrieve specific values for tests depending on the source
+// and destination data types. May be used to retrieve converted reference data.
+// All provided methods are save to use and protected from UB when call
+// static_cast<int>(unsigned int).
+template <typename SrcT, typename DstT> struct value_conv {
+  static SrcT lowest() {
+    if constexpr (!sts::is_unsigned_v<SrcT> && sts::is_unsigned_v<DstT>) {
+      if constexpr (sizeof(SrcT) > sizeof(DstT)) {
+        return static_cast<SrcT>(value<DstT>::lowest());
+      } else {
+        return value<SrcT>::lowest();
+      }
+    } else {
+      return std::max(value<SrcT>::lowest(),
+                      static_cast<SrcT>(value<DstT>::lowest()));
+    }
+  }
+
+  static SrcT max() {
+    if constexpr (!sts::is_unsigned_v<SrcT> && sts::is_unsigned_v<DstT>) {
+      if constexpr (sizeof(SrcT) > sizeof(DstT)) {
+        return static_cast<SrcT>(value<DstT>::max());
+      } else {
+        return value<SrcT>::max();
+      }
+    } else {
+      return std::min(value<SrcT>::max(),
+                      static_cast<SrcT>(value<DstT>::max()));
+    }
+  }
+
+  static SrcT denorm_min() {
+    if constexpr (std::is_same_v<SrcT, sycl::half> ||
+                  std::is_same_v<DstT, sycl::half>) {
+      return value<SrcT>::denorm_min();
+    } else if constexpr (sizeof(SrcT) > sizeof(DstT)) {
+      return static_cast<SrcT>(value<DstT>::max());
+    } else {
+      return value<SrcT>::max();
+    }
```

Comment:
```suggestion
    if constexpr (!type_traits::is_sycl_floating_point_v<DataT>) {
      // Return zero for any integral type the same way std::denorm_min does
      return 0;
    } else if constexpr (sizeof(SrcT) > sizeof(DstT)) {
      // Use the biggest value so it would not degenerate to zero during conversion
      // from SrcT to DstT
      return static_cast<SrcT>(value<DstT>::denorm_min());
    } else {
      return value<SrcT>::denorm_min();
    }
```

So, we need to have something like `std::max(src_denorm, dst_denorm)` with UB resolved, right?

1. May be it's better to state explicitly value returned for non-floating point `SrcT` case?
2. Why do we need to return `value<SrcT>::denorm_min()` for the case of `using DstT = sycl::half`? By doing so we return the minimal value out of the DstT and SrcT values.
3. According to SYCL 2020 "The half data type must conform to the IEEE 754-2008 half precision storage format.". It means we don't need the separate branches for the `sycl::half` at all. Am I missing something?

Reply:
I've applied this comment in this commit: 5959fb1f2f4dd5aca794998f3a625824bfc42107